tools/gbcpal.c

```diff
@@ -0,0 +1,156 @@
+	return (struct Color){
+		.r = lo & 0x1f,
+		.g = ((hi << 3) & 0x18) | ((lo >> 5) & 0x07),
+		.b = (hi >> 2) & 0x1f,
+	};
+}
+
+double luminance(struct Color color) {
+	return 0.299 * color.r * color.r + 0.587 * color.g * color.g + 0.114 * color.b * color.b;
+}
+
+int compare_colors(const void *color1, const void *color2) {
+	double lum1 = luminance(*(const struct Color *)color1);
+	double lum2 = luminance(*(const struct Color *)color2);
+	return lum1 < lum2 ? 1 : lum1 > lum2 ? -1 : 0;
+}
+
+void read_gbcpal(const char *filename, struct Color **colors, size_t *num_colors) {
+	long filesize;
+	uint8_t *bytes = read_u8(filename, &filesize);
+	if (filesize == 0) {
+		error_exit("%s: empty gbcpal file\n", filename);
+	}
+	if (filesize % 2) {
+		error_exit("%s: invalid gbcpal file\n", filename);
+	}
+
+	size_t new_colors = filesize / 2;
+	*colors = xrealloc(*colors, (sizeof **colors) * (*num_colors + new_colors));
+	for (size_t i = 0; i < new_colors; i++) {
+		(*colors)[*num_colors + i] = unpack_color(bytes[i * 2], bytes[i * 2 + 1]);
+	}
+	*num_colors += new_colors;
+
+	free(bytes);
+}
+
+void filter_colors(struct Color **colors, size_t *num_colors) {
+	size_t num_filtered = 0;
+	struct Color *filtered = xmalloc((sizeof **colors) * *num_colors);
+
+	for (size_t i = 0; i < *num_colors; i++) {
+		struct Color color = (*colors)[i];
+		if (color.r == 0 && color.g == 0 && color.b == 0) {
+			continue; // filter out black
+		}
+		if (color.r == 31 && color.g == 31 && color.b == 31) {
+			continue; // filter out white
+		}
+		if (num_filtered > 0) {
+			struct Color last = filtered[num_filtered - 1];
+			if (color.r == last.r && color.g == last.g && color.b == last.b) {
+				continue; // filter out duplicates
+			}
+		}
+		filtered[num_filtered++] = color;
+	}
+
+	free(*colors);
+	*num_colors = num_filtered;
+	*colors = filtered;
+}
+
+int main(int argc, char *argv[]) {
+	parse_args(argc, argv);
+
+	argc -= optind;
+	argv += optind;
+	if (argc < 1) {
+		usage_exit(1);
+	}
+
+	const char *out_filename = argv[0];
+
+	struct Color *colors = NULL;
+	size_t num_colors = 0;
+	if (argc == 1) {
+		read_gbcpal(out_filename, &colors, &num_colors);
+	}
+	for (int i = 1; i < argc; i++) {
+		read_gbcpal(argv[i], &colors, &num_colors);
+	}
+
+	qsort(colors, num_colors, sizeof(*colors), compare_colors);
+	filter_colors(&colors, &num_colors);
+
+	uint16_t first, second;
+	if (num_colors == 0) {
+		// colors are white and black
+		first = 0x7fff;
+		second = 0x0000;
+	} else if (num_colors == 1) {
+		// colors are both the same one
+		first = pack_color(colors[0]);
+		second = first;
+	} else if (num_colors == 2) {
+		// colors are light and dark
+		first = pack_color(colors[0]);
+		second = pack_color(colors[1]);
+		if (reverse) {
+			// darker color is first
+			uint16_t tmp = first;
+			first = second;
+			second = tmp;
+		}
+	} else {
+		error_exit("%s: more than 2 colors besides black and white (%zu)\n", out_filename, num_colors);
+	}
```

Comment:
I think the algorithm of `white + reversed?(colors) + black * (1 + (max_colors - num_colors))` would make this cleaner, something like
```c
if (num_colors > (MAX_COLORS - 2)) error;

uint16_t colors[MAX_COLORS];
int c = 0;
colors[c++] = 0x7fff;  // First color is white
for (int i = 0; i < num_colors; i++) colors[c++] = pack_color(colors[reverse ? num_colors - 1 - i : i]);  // Add the maximum of 2 custom colors
while (c < MAX_COLORS / 2) colors[c++] = 0x7fff;  // If not enough colors are present, fill up to half with white
while (c < MAX_COLORS) colors[c++] = 0;  // Fill the remainder with black

// Alternatively: fill the remaining colors from a predefined grayscale array, to make battle anims nicer?
for (c < MAX_COLORS; c++) colors[c] = pack_color((struct Color){32/4*(3-c), 32/4*(3-c), 32/4*(3-c)});
```

This is less verbose and I like the option of expanding this for more colors/maybe gen3 in the future?

Reply:
I find this a lot harder to understand and grasp "when my input was such-and-such, what order will my colors be in?" I think the current verbose code is useful for being able to point users to it and show just how this new approach works.

Extending it to support 4bpp is a cool idea, but that would involve more rewriting than just this section, and it's a short enough file that I don't think we need to future-proof it like that already. (YAGNI; a gbapal.c tool will be its own thing, not much harder to write than this was.)

Reply:
> fill the remaining colors from a predefined grayscale array

I did consider this, but it would be error-prone. Imagine a front sprite that uses one or two colors, and a back sprite using zero, just black+white. Each one gets gbapal.c run on it, so the back sprite now has two gray colors... and then generating normal.gbcpal fails because there are too many combined colors.